protocol/app/process/slinky_market_price_decoder.go

```diff
@@ -0,0 +1,120 @@
+package process
+
+import (
+	"fmt"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dydxprotocol/v4-chain/protocol/app/prepare/prices"
+	pricestypes "github.com/dydxprotocol/v4-chain/protocol/x/prices/types"
+	slinkyabci "github.com/skip-mev/slinky/abci/types"
+	"github.com/skip-mev/slinky/abci/ve"
+)
+
+// SlinkyMarketPriceDecoder wraps an existing MarketPriceDecoder with logic to verify that the MarketPriceUpdateTx
+// was indeed derived from vote-extensions injected into the block.
+type SlinkyMarketPriceDecoder struct {
+	// underlying UpdateMarketPriceTxDecoder
+	decoder UpdateMarketPriceTxDecoder
+
+	// underlying vote-extension aggregator
+	agg prices.PriceUpdateGenerator
+}
+
+// NewSlinkyMarketPriceDecoder returns a new SlinkyMarketPriceDecoder
+func NewSlinkyMarketPriceDecoder(
+	decoder UpdateMarketPriceTxDecoder, agg prices.PriceUpdateGenerator) *SlinkyMarketPriceDecoder {
+	return &SlinkyMarketPriceDecoder{
+		decoder: decoder,
+		agg:     agg,
+	}
+}
+
+// DecodeUpdateMarketPricesTx returns a new `UpdateMarketPricesTx` after validating the following:
+//   - the underlying decoder decodes successfully
+//   - the UpdateMarketPricesTx follows correctly from the vote-extensions
+//   - vote-extensions are enabled: each price per market-id is derived from the injected extended commit
+//   - vote-extensions are disabled: no price updates are proposed
+//   - vote-extensions are inserted into the block if necessary
+//
+// If error occurs during any of the checks, returns error.
+func (mpd *SlinkyMarketPriceDecoder) DecodeUpdateMarketPricesTx(
+	ctx sdk.Context, txs [][]byte) (*UpdateMarketPricesTx, error) {
+	expectedMsg := &pricestypes.MsgUpdateMarketPrices{}
+
+	// check if vote-extensions are enabled
+	if ve.VoteExtensionsEnabled(ctx) {
+		// if there isn't a vote-extension in the block when there should be, fail
+		if len(txs) < slinkyabci.NumInjectedTxs {
+			return nil, getDecodingError(
+				msgUpdateMarketPricesType,
+				fmt.Errorf("expected %v txs, got %v", slinkyabci.NumInjectedTxs, len(txs)))
+		}
+
+		// get the expected message from the injected vote-extensions
+		var err error
+		expectedMsg, err = mpd.agg.GetValidMarketPriceUpdates(ctx, txs[slinkyabci.OracleInfoIndex])
+		if err != nil {
+			return nil, getDecodingError(msgUpdateMarketPricesType, err)
+		}
+	}
+
+	// use the underlying decoder to get the UpdateMarketPricesTx
+	updateMarketPrices, err := mpd.decoder.DecodeUpdateMarketPricesTx(ctx, txs)
```

Comment:
what is the purpose of this inner decoder? Is it so `SlinkyMarketPriceDecoder` can call `DefaultMarketPriceDecoder`?

Reply:
@nivasan1 can probably comment more on this as he implemented it.

The idea is twofold:
1. The existing behavior is preserved via just using `DefaultMarketPricesDecoder` so all of these changes are backwards compatible in the case where slinky is not yet enabled.
2. The new implementation wraps the default encoder because we still want all the same behavior. But additionally we want to do checks on the equality of VE data with the changes being made in the UpdateMarketPrices tx. 

So we first get the injected VE data, then get the UpdateMarketPricesTx via the default behavior, then verify that the two match.

Reply:
That's correct. The `SlinkyUpdateMarketPriceDecoder`'s responsibility is to decode the price-update + determine whether or not the proposed prices followed exactly from the injected vote-extensions